### DISCUSSION TITLE
Ingestion Pipeline

### DISCUSSION THREAD

**Discussion by CrossBound** [datalust/seq-tickets#1218] [Ideas]
Maybe this is asking too much (but then again... Seq is one of the best quality products I've seen) but it would be fantastic if it were possible to have a data ingestion pipeline which would allow custom middleware components to  filter/operate on incoming logs in sequence. In concept this is identical to the way ASP.NET Core handles HTTP requests, but this would be for operating on logs. Basically at the start of the pipeline would be the input (native API or custom input app), then one or more middleware components in an ordered sequence, then finally the logs land in the log stream. 

Right now Seq already supports custom apps as a _source_  of logs, but once the logs are created by the app there is not a hook to be able to support custom processes. Example usages might be attaching custom properties based on the event (not currently possible), filtering based on the data (possible if you can make it an expression), and transforming the data before it reaches the stream. 

In fact, combined with the ability to install middleware from nuget (like we can install input apps from nuget) middleware would enable virtually any data source to be pointed at Seq and then turned into structured logging through middleware apps.

Use Case 1:
Right now we can ingest Syslog via the Seq.Input.Syslog package, but any custom data in the message body is not treated as structured data so it requires text processing (a performance hit I assume) to query it. A middleware component could access the event data & properties and parse out the necessary fields. I know issue https://github.com/datalust/squiflog/issues/23 already suggests one possible option, but it has a very focused solution. Middleware would enable many additional scenarios. 

Use Case 2:
An incoming stream needs to have a different property attached based on the data, Going back to the Syslog input example, we have a WAF outputting logs for multiple applications that are behind it. We would need a way to sort them into different buckets (in my case we tag each log with a "System" property which identifies the application) based on which application the log is for. Currently this is only possible by running multiple instances of the Syslog input app with different filters and the same stream being replicated to all of them (very impractical especially with a high volume input stream).

**Comment by nblumhardt** [comment]
Thanks, @CrossBound - we've had similar thoughts recently and will explore this further 👍

**Comment by nblumhardt** [comment]
As part of #1126 we're doing some additional work on how app-based inputs are configured. We won't be able to add support for middleware pipelines there yet, but we'll keep this front-of-mind and aim to set things up in a way that will make this easier in the future.

If you're keen to explore what this kind of middleware would look like, it's very likely it would receive configuration via environment variables (as apps now do), accept strict CLEF-formatted, newline-delimited JSON events on `STDIN`, and write the same format to `STDOUT`.

You could build and run this kind of thing today, if you run it manually, by writing your middleware as an EXE and using `seqcli` to ingest the final result.

Assuming you've unpacked _Seq.Input.Syslog_, it would be something like:

```powershell
$env:SEQ_APP_SETTING_SYSLOGADDRESS = "udp://0.0.0.0:514"

& ./Seq.Input.Syslog/win-x64/squiflog.exe `
  | ./YourMiddleware/YourMiddleware.exe `
  | seqcli ingest --json
```

(Untested, but hopefully illustrates the idea.)

Depending on what kind of ops environment you use, it might be possible to just run this as an additional process alongside Seq (assigning a unique API key so that you can control it via _Inputs_), or package it up as a Docker container to run as a sidecar.

Keen to know how this matches up with what you expect, and let me know if you decide to give it a shot!

Best regards,
Nick

**Reply by CrossBound** [reply]
That looks like it would do the job... the only down-side I'm seeing is the fact that since you require json in and json out that means each app in the pipeline will have to parse the same json over and over... a performance impact. Otherwise, I think that will work, and yes, I believe it would solve the use cases mentioned.

**Comment by PassivePicasso** [comment]
I'm not sure if this is on the same track. I've got a situation where I'm logging data from some vendor libraries.  Those libraries have their logging implemented via Splat, however they don't follow good rules for message templates, leading to template bloat on Seq.NET

I've tried to work with the library maintainers to resolve this issue, but they are not interested in fixing these inconsistencies, nor accepting PRs to fix them. As a result I have to disable the logging I have for these libraries. 

These logs pass through my standard logging system in my client application, which uses Serilog for logging and adheres to Seq.NET best practices as well as I can manage, thus the logs come over the HTTP end point on Seq.NET.

While I could implement handling in the client that would modify the logs before being sent out, it would be more convenient to be able to do this from the Seq.NET interface to create expressions that evaluate logs then convert them into structured logs entries.

**Reply by nblumhardt** [reply]
Thanks for the note, that's good info to have - I think it's a valid use case for what we're discussing in this thread 👍 

If you need a hand figuring out how to do this client-side with Serilog in the meantime, please just let me know.

**Reply by PassivePicasso** [reply]
Would a Seq App be capable of handling these modifications?

**Reply by nblumhardt** [reply]
@PassivePicasso not while hosted in Seq, currently - apps can act as inputs, but to do this the app would need to receive the events (from the API) somehow first.
